### PR TITLE
'main': Highlight the ! precommand as an error when not at the start of a pipeline.

### LINKED ISSUE #511: Incorrectly placed ! not highlighted as syntax error
In `echo | ! cat` `!` is highlighted as `reserved-word` but should be highlighted as `unknown-token` as it's a syntax error (since `!` can only occur at the start of a pipeline).

### PR DESCRIPTION
**Pull request by danielshahaf**
Fixes #511.

### CODE REVIEW

File: highlighters/main/main-highlighter.zsh [hunk 1]
```diff
@@ -822,15 +833,15 @@ _zsh_highlight_main_highlighter_highlight_list()
...
+      next_word=':start::start_of_pipeline:'
      fi
    else # $arg is a non-command word
       case $arg in
         $'\x29') # subshell or end of array assignment
                  if $in_array_assignment; then
                    style=assign
                    in_array_assignment=false
-                   next_word+=':start:'
+                   next_word+=':start::start_of_pipeline:'
```
Comment (phy1729) [inline highlighters/main/main-highlighter.zsh]: `foo=(bar) ! false` is a parse error here, so I think this is incorrect.

File: highlighters/main/main-highlighter.zsh [hunk 2]
```diff
@@ -684,6 +689,9 @@ _zsh_highlight_main_highlighter_highlight_list()
       case $res in
         reserved)       # reserved word
                         style=reserved-word
+                        if [[ $arg == '!' && $this_word != *':start_of_pipeline:'* ]]; then
```
[the remark this answers is not in the document]
Comment (danielshahaf) [reply highlighters/main/main-highlighter.zsh]: Changed.

File: highlighters/main/main-highlighter.zsh (hunk 1 above)
Comment (danielshahaf) [reply highlighters/main/main-highlighter.zsh]: Fixed.

**Comment by danielshahaf** [comment]
Added tests and addressed the comments.  Thanks for the review.

**Pull request merged**